app/src/main/java/com/github/epfl/meili/forum/FirebasePostService.kt

```diff
@@ -0,0 +1,75 @@
+package com.github.epfl.meili.forum
+
+import android.util.Log
+import com.github.epfl.meili.forum.Post.Companion.toPost
+import com.google.firebase.firestore.DocumentChange
+import com.google.firebase.firestore.FirebaseFirestore
+import java.util.*
+
+class FirebasePostService() : PostService() {
+
+    companion object {
+        val DEFAULT_DB = { FirebaseFirestore.getInstance() }
+
+        // Change this for dependency injection
+        var dbProvider: () -> FirebaseFirestore = DEFAULT_DB
+    }
+
+    private val TAG = "FirebasePostService"
+    private var db: FirebaseFirestore = dbProvider()
+
+    init {
+        // Listen to changes in post collection and make appropriate changes
+        db.collection("posts").addSnapshotListener { snapshot, e ->
+            // Handle errors
+            if (e != null || snapshot == null) {
+                Log.w(TAG, "Listen failed.", e)
+                return@addSnapshotListener
+            }
+
+            // Modify local list for each change in Firestore database
+            for (post in snapshot!!.documentChanges) {
+                when (post.type) {
+                    DocumentChange.Type.ADDED -> posts.add(post.document.toPost()!!)
+                    DocumentChange.Type.MODIFIED -> {
+                        posts.remove(post.document.toPost())
+                        posts.add(post.document.toPost()!!)
+                    }
+                    DocumentChange.Type.REMOVED -> posts.remove(post.document.toPost())
+                }
+            }
```

Comment:
Is refactoring here like this possible or is this incorrect?
```suggestion
            for (post in snapshot!!.documentChanges) {
                val postDocument = post.document.toPost()!!
                when (post.type) {
                    DocumentChange.Type.ADDED -> posts.add(postDocument )
                    DocumentChange.Type.MODIFIED -> {
                        posts.remove(postDocument )
                        posts.add(postDocument )
                    }
                    DocumentChange.Type.REMOVED -> posts.remove(postDocument )
                }
            }
```

Reply:
It is incorrect, I need to make Posts out of the information stored in the database via toPost()

Reply:
Like I do not store the posts directly, only the relevant information